Testing test_components_from_metric_edge_cases breaks around line 130 on some environments

This might be a particularly weird and not primordial issue.

#### Description
One of the tests fails on a newly created environment, depending on the platform I test on.

#### Steps/Code to Reproduce
With Python 3.7.5 installed with Anaconda:
```bash
conda create -n test_env numpy scipy pytest scikit-learn
conda activate test_env
pytest test/test_components_metric_conversion.py
```

#### Expected Results
Tests run without failure on any platform.

#### Actual Results
Fails around line 130 of `test/test_components_metric_conversion.py` for on the two platform tested.
```
============================================== FAILURES 
[...]
      assert np.linalg.slogdet(M)[1] > 1  # (just to show that the computed
      # determinant is far from null)
      with pytest.raises(LinAlgError) as err_msg:
>       np.linalg.cholesky(M)
E       Failed: DID NOT RAISE <class 'numpy.linalg.LinAlgError'>

test/test_components_metric_conversion.py:130: Failed
```

#### Versions
The environment in which it fails:
```
Linux-4.18.0-25-generic-x86_64-with-debian-buster-sid
Python 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
NumPy 1.17.2
SciPy 1.3.1
Scikit-Learn 0.21.3
Metric-Learn 0.5.0
```
The environment in which it passes:
```
Linux-5.0.0-32-generic-x86_64-with-debian-buster-sid
Python 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
NumPy 1.17.2
SciPy 1.3.1
Scikit-Learn 0.21.3
Metric-Learn 0.5.0
```
So the only difference is the platform.

--- a/test/test_components_metric_conversion.py
+++ b/test/test_components_metric_conversion.py
@@ -1,7 +1,6 @@
 import unittest
 import numpy as np
 import pytest
-from numpy.linalg import LinAlgError
 from scipy.stats import ortho_group
 from sklearn.datasets import load_iris
 from numpy.testing import assert_array_almost_equal, assert_allclose

--- a/test/test_components_metric_conversion.py
+++ b/test/test_components_metric_conversion.py
@@ -117,17 +117,14 @@ class TestTransformerMetricConversion(unittest.TestCase):
     L = components_from_metric(M)
     assert_allclose(L.T.dot(L), M)
 
-    # matrix with a determinant still high but which should be considered as a
-    # non-definite matrix (to check we don't test the definiteness with the
-    # determinant which is a bad strategy)
+    # matrix with a determinant still high but which is
+    # undefinite w.r.t to numpy standards
     M = np.diag([1e5, 1e5, 1e5, 1e5, 1e5, 1e5, 1e-20])
     M = P.dot(M).dot(P.T)
     assert np.abs(np.linalg.det(M)) > 10
     assert np.linalg.slogdet(M)[1] > 1  # (just to show that the computed
     # determinant is far from null)
-    with pytest.raises(LinAlgError) as err_msg:
-      np.linalg.cholesky(M)
-    assert str(err_msg.value) == 'Matrix is not positive definite'
+    assert np.linalg.matrix_rank(M) < M.shape[0]
     # (just to show that this case is indeed considered by numpy as an
     # indefinite case)
     L = components_from_metric(M)

--- a/test/test_utils.py
+++ b/test/test_utils.py
@@ -1079,7 +1079,7 @@ def test__check_sdp_from_eigen_tol_non_psd(w, tol):
 
 @pytest.mark.unit
 @pytest.mark.parametrize('w, is_definite',
-                         [(np.array([1e5, 1e5, 1e5, 1e5, 
+                         [(np.array([1e5, 1e5, 1e5, 1e5,
                                      1e5, 1e5, 1e-20]), False),
                           (np.array([1e-10, 1e-10]), True)])
 def test__check_sdp_from_eigen_tol_default_psd(w, is_definite):

--- a/test/test_utils.py
+++ b/test/test_utils.py
@@ -1055,6 +1055,53 @@ def test__check_sdp_from_eigen_returns_definiteness(w, is_definite):
   assert _check_sdp_from_eigen(w) == is_definite
 
 
+@pytest.mark.unit
+@pytest.mark.parametrize('w, tol, is_definite',
+                         [(np.array([5., 3.]), 2, True),
+                          (np.array([5., 1.]), 2, False),
+                          (np.array([5., -1.]), 2, False)])
+def test__check_sdp_from_eigen_tol_psd(w, tol, is_definite):
+  """Tests that _check_sdp_from_eigen, for PSD matrices, returns
+  False if an eigenvalue is lower than tol"""
+  assert _check_sdp_from_eigen(w, tol=tol) == is_definite
+
+
+@pytest.mark.unit
+@pytest.mark.parametrize('w, tol',
+                         [(np.array([5., -3.]), 2),
+                          (np.array([1., -3.]), 2)])
+def test__check_sdp_from_eigen_tol_non_psd(w, tol):
+  """Tests that _check_sdp_from_eigen raises a NonPSDError
+  when there is a negative value with abs value higher than tol"""
+  with pytest.raises(NonPSDError):
+    _check_sdp_from_eigen(w, tol=tol)
+
+
+@pytest.mark.unit
+@pytest.mark.parametrize('w, is_definite',
+                         [(np.array([1e5, 1e5, 1e5, 1e5, 
+                                     1e5, 1e5, 1e-20]), False),
+                          (np.array([1e-10, 1e-10]), True)])
+def test__check_sdp_from_eigen_tol_default_psd(w, is_definite):
+  """Tests that the default tol argument gives good results for edge cases
+  like even if the determinant is high but clearly one eigenvalue is low,
+  (undefinite so returns False) or when all eigenvalues are low (definite so
+  returns True)"""
+  assert _check_sdp_from_eigen(w, tol=None) == is_definite
+
+
+@pytest.mark.unit
+@pytest.mark.parametrize('w',
+                         [np.array([1., -1.]),
+                          np.array([-1e-10, 1e-10])])
+def test__check_sdp_from_eigen_tol_default_non_psd(w):
+  """Tests that the default tol argument is good for raising
+  NonPSDError, e.g. that when a value is clearly relatively
+  negative it raises such an error"""
+  with pytest.raises(NonPSDError):
+    _check_sdp_from_eigen(w, tol=None)
+
+
 def test__check_n_components():
   """Checks that n_components returns what is expected
   (including the errors)"""

